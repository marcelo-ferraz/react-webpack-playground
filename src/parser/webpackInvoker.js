import * as Babel from '@babel/standalone';
import { find } from 'lodash';
import { resolve as pathResolve, dirname, basename, extname } from 'path';
import { js, json, jsx } from '../fileExtensions';
import { isItMeaningful } from './helpers';
import { resolve } from './helpers';

const findByPath = (obj, path) =>
    find(
        Object.entries(obj),
        ([key]) => key === path || key === `${path}.js` || key === `${path}/index.js`,
    ) || [];

const emptyFunc = function () {
    // eslint-disable-next-line no-useless-return
    return;
};

function reduceAllImports(code, whenFound, initialState) {
    const findImports = /__customRequire\(\\?("|')(?<importId>\.\.?\/[\w-\.\s\/]+)\\?("|')/gm;
    let acc = initialState;
    let match;
    do {
        match = findImports.exec(code);

        if (match && match.groups) {
            const { importId } = match.groups;
            acc = whenFound(acc, importId);
        }
    } while (match);
    return acc;
}

const parse = (rawCode) => {
    const { code } = Babel.transform(rawCode, { presets: ['es2015', 'react'] });
    return code.replace(/require\(/gm, '__customRequire(');
};

const customRequire = (path) => {
    const [keyFound] = findByPath(allDependencies, path);

    if (!keyFound) {
        /* eslint-disable-next-line no-undef */
        return __webpack_require__(resolve(path));
    }

    const { exports } = jsInvoke(dependencies[keyFound]);

    return exports;
};

export async function render(entries, entryPath = null) {
    let allDependencies = {};

    // eslint-disable-next-line no-unused-vars, no-underscore-dangle
    const __customRequire = (path) => {
        const [keyFound] = findByPath(allDependencies, path);

        if (!keyFound) {
            /* eslint-disable-next-line no-undef */
            return __webpack_require__(resolve(path));
        }

        const { exports } = jsInvoke(allDependencies[keyFound]);

        return exports;
    };

    const renderImpl = (path, dependencies = {}) => {
        const [, rawCode] = findByPath(entries, path);
        const unit = { i: `dynamic:${path}`, l: false, exports: {} };

        if (!isItMeaningful(rawCode)) {
            return { unit, dependencies };
        }

        const type = extname(path);

        switch (type) {
            case '':
            case js:
            case jsx: {
                const code = parse(rawCode);
                dependencies = reduceAllImports(
                    code,
                    (deps, id) => (!deps[id] ? { ...deps, [id]: renderImpl(id, deps) } : deps),
                    dependencies,
                );

                // eslint-disable-next-line no-eval
                const func = eval(`(function(module, exports, __webpack_require__) { ${code} })`);
                return { func, unit, dependencies };
            }
            case json: {
                // debugger;
                unit.exports = JSON.parse(rawCode);
                return { unit };
            }
            default:
                throw new TypeError(`${type} is not supported`);
        }
    };

    // render the tree starting from the root (main.js)
    const context = renderImpl(entryPath || './app.js');
    allDependencies = context.dependencies;
    return context;
}

export function jsInvoke({ func, unit }) {
    if (func) {
        // eslint-disable-next-line no-undef
        func.call(unit.exports, unit, unit.exports, __webpack_require__);
    }

    return unit;
}
