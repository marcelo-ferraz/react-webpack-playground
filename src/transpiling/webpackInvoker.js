import { js, json, jsx } from '../fileExtensions';
import { pathFinder } from './helpers';
import { tryResolve } from './helpers';
import MissingModuleError from './errors/MissingModuleError';
import parseNShake from './parseNShake';

// eslint-disable-next-line no-undef
const customRequireImpl = function (use = __webpack_require__, dependencies, path) {
    let fullPath = tryResolve(dependencies, path);

    if (fullPath) {
        const dependency = dependencies[fullPath];

        if (dependency.__esModule) {
            return dependency;
        }

        const invoked = jsInvoke(dependency);
        return invoked.exports || invoked;
    }

    fullPath = tryResolve(use.m, path);
    if (fullPath) {
        return use(fullPath);
    }

    try {
        return use(path);
    } catch (error) {
        throw new MissingModuleError(path, error);
    }
};

const loadAllExtraDependencies = async ({ getExtraDependencies }) => {
    if (!getExtraDependencies) {
        return {};
    }

    const dependencies = getExtraDependencies();
    const dependencyValues = Promise.all(Object.values(dependencies));
    const dependencyKeys = Object.keys(dependencies);

    return (await dependencyValues).reduce((deps, depValue, i) => {
        return { ...deps, [dependencyKeys[i]]: depValue };
    }, {});
};

let parserWorker;
const defaultEntryPath = './App';

const getParserWorker = () => {
    if (!parserWorker) {
        if (!window.Worker) {
            throw new Error('The browser doesnt support web workers');
        }

        parserWorker = new Worker(new URL('../workers/parserWorker', import.meta.url));
    }

    return parserWorker;
};

async function renderHere(strategy, entryPath, finder) {
    const jsFinder = finder || pathFinder(strategy.entries);

    const es5Entries = parseNShake(jsFinder, entryPath || defaultEntryPath);

    return await renderImpl(es5Entries, entryPath, strategy);
}

async function renderElsewhere(strategy, entryPath = defaultEntryPath) {
    const worker = getParserWorker();
    worker.postMessage([strategy.entries, entryPath]);

    return new Promise((resolve, reject) => {
        worker.onmessage = async ({ data: entries }) => {
            try {
                resolve(await renderImpl(entries, entryPath, strategy));
            } catch (e) {
                reject(e);
            }
        };
    });
}

function jsInvoke(context) {
    if (context.func) {
        if (context.events && context.events.beforeInvoke) {
            context.events.beforeInvoke(context);
        }

        context.func(context.unit, context.unit.exports, __webpack_require__);

        if (context.events && context.events.afterInvoke) {
            context.events.afterInvoke(context);
        }
    }
    return context.unit;
}

async function renderImpl(entries, entryPath, strategy) {
    let dynamicContext = null;
    let dependencies = {};

    // this custom require is the entry point to the module system compiled in dev mode
    // eslint-disable-next-line no-unused-vars, no-underscore-dangle
    function __customRequire(req, path) {
        return customRequireImpl(req, dependencies, path);
    }

    let extraDependencies = (await loadAllExtraDependencies(strategy)) || {};

    if (strategy.beforeRender) {
        extraDependencies = strategy.beforeRender(strategy, extraDependencies);
    }

    dependencies = { ...dependencies, ...extraDependencies };

    for (const [path, entry] of Object.entries(entries)) {
        const { ext, code } = entry;
        const unit = { i: `dynamic:${path}`, l: false, exports: {} };

        if (!code) {
            continue;
        }

        switch (ext) {
            case '':
            case js:
            case jsx: {
                // eslint-disable-next-line no-eval
                const func = eval(code);

                const ctx = {
                    func,
                    unit,
                    customRequire: __customRequire,
                };

                if (path === entryPath) {
                    dynamicContext = ctx;
                }

                const id = path.endsWith(ext) ? path.substring(0, path.length - ext.length) : path;

                dependencies[id] = ctx;
                break;
            }
            case json: {
                unit.exports = code;
                dependencies[path] = { unit };
                break;
            }
        }
    }

    if (dynamicContext && dynamicContext.func) {
        dynamicContext.customRequire = (path) =>
            customRequireImpl(__webpack_require__, dependencies, path);
        dynamicContext.addDependencies = (key, value) => {
            dependencies = { ...dependencies, [key]: value };
        };

        dynamicContext.events = {
            beforeRender: strategy.beforeRender,
            beforeInvoke: strategy.beforeInvoke,
            afterInvoke: strategy.afterInvoke,
        };

        return dynamicContext;
    }
}

export { renderHere, renderElsewhere, jsInvoke, defaultEntryPath };
