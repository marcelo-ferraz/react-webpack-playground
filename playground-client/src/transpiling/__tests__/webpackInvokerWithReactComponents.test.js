import React from 'react';
import { render as testRender } from '@testing-library/react';
import cases from 'jest-in-case';
import simpleReactComponentMock from '../__mocks__/simpleReactComponent.mock';
import twoReactComponentsMock from '../__mocks__/twoReactComponents.mock';
import twoReactComponentsNJsonMock from '../__mocks__/twoReactComponentsNJson.mock';
import { jsInvoke, render } from '../webpackInvoker';

/* eslint-disable no-undef */
global.__webpack_require__ = function (id) {
    if (id && id.indexOf('react') >= 0) {
        return React;
    }
};
global.__webpack_require__.m = [{ './node_modules/react': {} }];
/* eslint-enable */

cases(
    'webpack invoker happy path for react',
    async ({ entry }) => {
        const { exports } = jsInvoke(await render(entry));

        const Component = exports.default;

        const { container } = testRender(<Component />);

        expect(container).toMatchSnapshot();
    },
    {
        'simple component': { entry: simpleReactComponentMock },
        'two components': { entry: twoReactComponentsMock },
        'two components and json': { entry: twoReactComponentsNJsonMock },
    },
);
