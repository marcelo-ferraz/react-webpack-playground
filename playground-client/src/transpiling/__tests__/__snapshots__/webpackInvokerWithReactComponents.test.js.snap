// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webpack invoker happy path for react simple component 1`] = `
<div>
  <div
    class="App"
  >
    <h1>
      Hello There
    </h1>
    <h2>
      Start the magic!
    </h2>
  </div>
</div>
`;

exports[`webpack invoker happy path for react two components 1`] = `
<div>
  <div
    class="App"
  >
    <h1>
      Hello CodeSandbox
    </h1>
    <h2>
      Start editing to see some magic happen!
    </h2>
    <div>
      here, some more magic
    </div>
  </div>
</div>
`;

exports[`webpack invoker happy path for react two components and json 1`] = `
<div>
  <div
    class="App"
  >
    <h1>
      Hello CodeSandbox
    </h1>
    <h2>
      Start editing to see some magic happen!
    </h2>
    <div>
      here, some more magic,
    </div>
    <div>
      and here some more of the same json: 
      {"id":42,"hash":"82d06ce0-e19d-4942-853a-4c74d4ae883e"}
    </div>
    <div>
      some json: 
      {"id":42,"hash":"82d06ce0-e19d-4942-853a-4c74d4ae883e"}
    </div>
  </div>
</div>
`;
